src/core/useAtom.ts

```diff
@@ -54,39 +54,49 @@ export function useAtom<Value, Update, Result extends void | Promise<void>>(
         throw atomState.p // read promise
       }
       if ('v' in atomState) {
-        return atomState.v as ResolveType<Value>
+        return [atomState.v, atomState.r] as [ResolveType<Value>, number]
       }
       throw new Error('no atom value')
     },
     [store, atom]
   )
 
   // Pull the atoms's state from the store into React state.
-  const [[version, value, atomFromUseReducer], rerenderIfChanged] = useReducer<
-    Reducer<
-      readonly [VersionObject | undefined, ResolveType<Value>, Atom<Value>],
-      VersionObject | undefined
-    >,
-    undefined
-  >(
-    useCallback(
-      (prev, nextVersion) => {
-        const nextValue = getAtomValue(nextVersion)
-        if (Object.is(prev[1], nextValue) && prev[2] === atom) {
-          return prev // bail out
-        }
-        return [nextVersion, nextValue, atom]
-      },
-      [getAtomValue, atom]
-    ),
-    undefined,
-    () => {
-      // NOTE should/could branch on mount?
-      const initialVersion = undefined
-      const initialValue = getAtomValue(initialVersion)
-      return [initialVersion, initialValue, atom]
-    }
-  )
+  const [[version, value, revision, atomFromUseReducer], rerenderIfChanged] =
+    useReducer<
+      Reducer<
+        readonly [
+          VersionObject | undefined,
+          ResolveType<Value>,
+          number,
+          Atom<Value>
+        ],
+        VersionObject | undefined
+      >,
+      undefined
+    >(
+      useCallback(
+        (prev, nextVersion) => {
+          const [nextValue, nextRevision] = getAtomValue(nextVersion)
+          if (
+            Object.is(prev[1], nextValue) &&
+            Object.is(prev[2], nextRevision) &&
```

Comment:
This technically means, when we have the same value, but a different revision, this hook triggers _unnecessary_ re-render. We should avoid it and should make a test case to catch it if possible. (guessing, it may or may not happen with suspense.)

Reply:
> This technically means, when we have the same value, but a different revision ...

That's true, but was exactly my case where the problem emerged: An atom that conditionally read another atom which should invoke `onMount` and start a subscription, but had initially the same default value.

> ... this hook triggers unnecessary re-render

 Because the same value was returned as in the previous render cycle, the effect was not getting re-rendered, therefore `commitAtom` not fired and `flushPending` not invoked.

With the current implementation the effect firing `commitAtom` should be a good place to call it.
I guess, as an alternative it could be also fired directly in the `rerenderIfChanged` dispatcher when the revision changed. This would save a render cycle.

```ts
      useCallback(
        (prev, nextVersion) => {
          const [nextValue, nextRevision] = getAtomValue(nextVersion)

          if (
            Object.is(prev[1], nextValue) &&
            Object.is(prev[2], nextRevision) &&
            prev[3] === atom
          ) {
            return prev // bail out
          }

          if (!Object.is(prev[2], nextRevision)) {
             setTimeout(() => store[COMMIT_ATOM](atom, version))
          }
          
          return [nextVersion, nextValue, nextRevision, atom]
        },
        [getAtomValue, atom]
      ),
```
https://github.com/pmndrs/jotai/blob/84a0d685b2f774b684673f488897a9f525e56e3d/src/core/useAtom.ts#L78-L91

Reply:
Please forgive me, I didn't looked your issue #947 closely yet. You filed this PR so quickly. 😅 

I think if we suspend and resolve with the same state, not useEffect will fire. Or maybe not?

Can you do some experiments without jotai only with React useState, and see if useEffect fires with the said situation?

In any case, this can be an issue on jotai end. If useEffect doesn't fire, and we don't want to do side effects in dispatch phase, so the only solution would be requiring `Provider` to fire effects. Okay, I feel like I need to sit down and take some time to think about it. If you can create some test cases, that would be helpful.

Reply:
Wait, this is not related to Suspense? My misunderstanding...